#Regresión logística

In [103]:
import pandas as pd
import numpy as np

#ETL
Importar el dataset

In [104]:
df = pd.read_csv("/content/data.csv")

In [105]:
df.head()

cliente_id     genero  edad pais    ciudad segmento_cliente  es_mayor  \
0  0002-ORFBO   Femenino  36.0  USA  New York             PYME         0   
1  0003-MKNFE  Masculino  45.0  USA  New York      Residencial         0   
2  0004-TLHLJ  Masculino  36.0  USA  New York             PYME         0   
3  0011-IGKFF  Masculino   NaN  USA  New York             PYME         1   
4  0013-EXCHZ   Femenino  48.0  USA  New York             PYME         1   

  tiene_pareja tiene_dependientes    latitud  ...  dias_mora  \
0           Si                 Si  40.726363  ...          0   
1           No                 No  40.678022  ...          0   
2           No                 No  40.638134  ...          0   
3           Si                 No  40.683321  ...          0   
4           Si                 No  40.764546  ...          0   

  cambio_plan_reciente  downgrade_reciente visitas_app_mensual  \
0                    1                   0                  66   
1                    0                   0                  51   
2                    0                   0                   5   
3                    0                   0                   9   
4                    0                   0                  37   

  tiempo_sesion_promedio  features_nuevas_usadas  ultimo_contacto_soporte  \
0                  11.51                       3               2025-11-27   
1                  20.17                       6               2025-11-10   
2                 109.44                       5               2025-10-29   
3                  90.94                       0               2025-11-14   
4                 113.28                       8               2025-08-12   

  competidores_area  ofertas_recibidas  precio_vs_mercado  
0                 4                  0        Competitivo  
1                 1                  2               Alto  
2                 6                  2        Competitivo  
3                 2                  3        Competitivo  
4                 1                  0               Alto  

[5 rows x 67 columns]

In [106]:
df.shape

(9701, 67)

In [107]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9701 entries, 0 to 9700
Data columns (total 67 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   cliente_id               9701 non-null   object 
 1   genero                   9701 non-null   object 
 2   edad                     8925 non-null   float64
 3   pais                     9701 non-null   object 
 4   ciudad                   9701 non-null   object 
 5   segmento_cliente         9701 non-null   object 
 6   es_mayor                 9701 non-null   int64  
 7   tiene_pareja             9701 non-null   object 
 8   tiene_dependientes       9701 non-null   object 
 9   latitud                  9701 non-null   float64
 10  longitud                 9701 non-null   float64
 11  borough                  9701 non-null   object 
 12  codigo_postal            9701 non-null   int64  
 13  estado                   9701 non-null   object 
 14  fecha_registro          

In [108]:
df.describe()

edad     es_mayor      latitud     longitud  codigo_postal  \
count  8925.000000  9701.000000  9701.000000  9701.000000    9701.000000   
mean     43.971092     0.161324    40.680135   -74.113888   10832.161324   
std      12.248348     0.367848     0.151019     0.769997     514.281652   
min      18.000000     0.000000    40.112385   -77.519855   10001.000000   
25%      35.000000     0.000000    40.634279   -73.997309   10312.000000   
50%      44.000000     0.000000    40.705993   -73.942995   11114.000000   
75%      53.000000     0.000000    40.758109   -73.877799   11239.000000   
max      74.000000     1.000000    40.910207   -73.701843   11696.000000   

       ingreso_mediano  densidad_poblacional   antiguedad  cargo_mensual  \
count      9701.000000           9701.000000  9701.000000    9701.000000   
mean      66460.674157          14736.604474    32.898876      89.254500   
std       14676.147445           7302.961070    25.605834      49.153898   
min       42000.000000           3200.000000     0.000000      15.000000   
25%       63000.000000           8500.000000     9.000000      48.280000   
50%       72000.000000          13000.000000    29.000000      90.620000   
75%       82000.000000          14000.000000    56.000000     127.660000   
max       85000.000000          28000.000000   179.000000     342.320124   

       ingresos_totales  ...  cancelacion  intentos_cobro_fallidos  \
count       9701.000000  ...  9701.000000              9701.000000   
mean        3208.115742  ...     0.156891                 0.996701   
std         3469.562390  ...     0.363717                 1.067799   
min            0.000000  ...     0.000000                 0.000000   
25%          488.040000  ...     0.000000                 0.000000   
50%         1705.550000  ...     0.000000                 1.000000   
75%         5155.840000  ...     0.000000                 2.000000   
max        29237.580000  ...     1.000000                 3.000000   

         dias_mora  cambio_plan_reciente  downgrade_reciente  \
count  9701.000000           9701.000000         9701.000000   
mean      7.632203              0.256365            0.213586   
std      10.032225              0.436648            0.409859   
min       0.000000              0.000000            0.000000   
25%       0.000000              0.000000            0.000000   
50%       5.000000              0.000000            0.000000   
75%      15.000000              1.000000            0.000000   
max      30.000000              1.000000            1.000000   

       visitas_app_mensual  tiempo_sesion_promedio  features_nuevas_usadas  \
count          9701.000000             9701.000000             9701.000000   
mean             49.405731               62.678872                4.984641   
std              28.884164               33.236751                3.153246   
min               0.000000                5.010000                0.000000   
25%              24.000000               34.230000                2.000000   
50%              50.000000               62.360000                5.000000   
75%              74.000000               91.690000                8.000000   
max              99.000000              119.990000               10.000000   

       competidores_area  ofertas_recibidas  
count        9701.000000        9701.000000  
mean            4.503762           0.981858  
std             2.284261           1.065524  
min             1.000000           0.000000  
25%             2.000000           0.000000  
50%             5.000000           1.000000  
75%             6.000000           2.000000  
max             8.000000           3.000000  

[8 rows x 36 columns]

In [109]:
df.columns

Index(['cliente_id', 'genero', 'edad', 'pais', 'ciudad', 'segmento_cliente',
       'es_mayor', 'tiene_pareja', 'tiene_dependientes', 'latitud', 'longitud',
       'borough', 'codigo_postal', 'estado', 'fecha_registro',
       'ingreso_mediano', 'densidad_poblacional', 'tipo_contrato',
       'antiguedad', 'cargo_mensual', 'ingresos_totales', 'metodo_pago',
       'canal_registro', 'errores_pago', 'descuento_aplicado',
       'aumento_precio_3m', 'facturacion_sin_papel', 'servicio_telefono',
       'lineas_multiples', 'tipo_internet', 'seguridad_online',
       'respaldo_online', 'proteccion_dispositivo', 'soporte_tecnico',
       'streaming_tv', 'streaming_peliculas', 'nivel_riesgo', 'score_riesgo',
       'conexiones_mensuales', 'dias_activos_semanales', 'promedio_conexion',
       'caracteristicas_usadas', 'tasa_crecimiento_uso',
       'dias_ultima_conexion', 'tickets_soporte', 'tiempo_resolucion',
       'tipo_queja', 'escaladas', 'puntuacion_csat', 'puntuacion_nps',
       'tasa_

Definir target y eliminar columnas innecesarias

In [110]:
TARGET = "cancelacion"

cols_to_drop = [
    "cliente_id",
    "fecha_registro",
    "fecha_ultimo_pago",
    "ultimo_contacto_soporte"
]

df = df.drop(columns=cols_to_drop, errors="ignore")

print(df.shape)

(9701, 63)


Separar X y y

In [111]:
X = df.drop(columns=[TARGET])
y = df[TARGET]

print(X.shape, y.shape)

(9701, 62) (9701,)


X train y X test

In [112]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.25,
    random_state=42,
    stratify=y
)

print("Train:", X_train.shape)
print("Test:", X_test.shape)

Train: (7275, 62)
Test: (2426, 62)


In [89]:
cat_cols = X_train.select_dtypes(include="object").columns.tolist()
num_cols = X_train.select_dtypes(exclude="object").columns.tolist()

print("Categóricas:", len(cat_cols))
print("Numéricas:", len(num_cols))

Categóricas: 27
Numéricas: 35


Preprocesamiento por imputación

In [113]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer

num_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

cat_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

preprocessor = ColumnTransformer(
    transformers=[
        ("num", num_pipeline, num_cols),
        ("cat", cat_pipeline, cat_cols)
    ]
)

#Modelo base

In [115]:
from sklearn.linear_model import LogisticRegression

model = Pipeline(steps=[
    ("preprocess", preprocessor),
    ("model", LogisticRegression(max_iter=1000, class_weight="balanced"))
])

model.fit(X_train, y_train)

Pipeline(steps=[('preprocess',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['edad', 'es_mayor',
                                                   'latitud', 'longitud',
                                                   'codigo_postal',
                                                   'ingreso_mediano',
                                                   'densidad_poblacional',
                                                   'antiguedad',
                                                   'cargo_mensual',
                                                   'ingresos_totales',
                                                   'errores_pago',
                                                   'score_riesgo',
                                                   'conexiones_men...
                                                   'aumento_precio_3m',
                                                   'facturacion_sin_papel',
                                                   'servicio_telefono',
                                                   'lineas_multiples',
                                                   'tipo_internet',
                                                   'seguridad_online',
                                                   'respaldo_online',
                                                   'proteccion_dispositivo',
                                                   'soporte_tecnico',
                                                   'streaming_tv',
                                                   'streaming_peliculas',
                                                   'nivel_riesgo', 'tipo_queja',
                                                   'respuesta_encuesta',
                                                   'precio_vs_mercado'])])),
                ('model',
                 LogisticRegression(class_weight='balanced', max_iter=1000))])

In [116]:
from sklearn.metrics import accuracy_score, recall_score, confusion_matrix

y_pred = model.predict(X_test)
y_proba = model.predict_proba(X_test)[:, 1]

In [117]:
from sklearn.metrics import accuracy_score, recall_score, confusion_matrix, roc_auc_score

acc = accuracy_score(y_test, y_pred)
rec = recall_score(y_test, y_pred)
auc = roc_auc_score(y_test, y_proba)
cm = confusion_matrix(y_test, y_pred)

print("Accuracy:", round(acc, 4))
print("Recall:", round(rec, 4))
print("ROC AUC:", round(auc, 4))
print("Matriz de confusión:\n", cm)

Accuracy: 0.9439
Recall: 0.8635
ROC AUC: 0.9476
Matriz de confusión:
 [[1961   84]
 [  52  329]]


#Ajuste del modelo

In [118]:
from sklearn.metrics import precision_score

thresholds = np.arange(0.05, 0.95, 0.05)

rows = []

for t in thresholds:
    y_pred_t = (y_proba >= t).astype(int)
    prec = precision_score(y_test, y_pred_t)
    rec = recall_score(y_test, y_pred_t)
    rows.append([t, prec, rec])

df_thr = pd.DataFrame(rows, columns=["threshold", "precision", "recall"])
df_thr

threshold  precision    recall
0        0.05   0.262465  0.994751
1        0.10   0.300733  0.968504
2        0.15   0.359517  0.937008
3        0.20   0.427696  0.916010
4        0.25   0.497110  0.902887
5        0.30   0.571669  0.889764
6        0.35   0.647287  0.876640
7        0.40   0.704255  0.868766
8        0.45   0.766204  0.868766
9        0.50   0.796610  0.863517
10       0.55   0.813433  0.858268
11       0.60   0.827411  0.855643
12       0.65   0.835897  0.855643
13       0.70   0.838046  0.855643
14       0.75   0.840206  0.855643
15       0.80   0.840206  0.855643
16       0.85   0.840206  0.855643
17       0.90   0.838961  0.847769

In [120]:
t = 0.30
y_pred_30 = (y_proba >= t).astype(int)

from sklearn.metrics import confusion_matrix, precision_score, recall_score

cm_30 = confusion_matrix(y_test, y_pred_30)
prec_30 = precision_score(y_test, y_pred_30)
rec_30 = recall_score(y_test, y_pred_30)

print("Threshold:", t)
print("Precision:", round(prec_30, 4))
print("Recall:", round(rec_30, 4))
print("Matriz de confusión:\n", cm_30)

Threshold: 0.3
Precision: 0.5717
Recall: 0.8898
Matriz de confusión:
 [[1791  254]
 [  42  339]]


Obtener probabilidades de churn

In [95]:
y_prob = model_base.predict_proba(X_test)[:, 1]

Curva Precision–Recall

In [96]:
from sklearn.metrics import precision_recall_curve

precision, recall, thresholds = precision_recall_curve(y_test, y_prob)

Tabla de trade-off

In [97]:
pr_df = pd.DataFrame({
    "threshold": thresholds,
    "precision": precision[:-1],
    "recall": recall[:-1]
})

pr_df.head()

threshold  precision  recall
0   0.000057   0.157049     1.0
1   0.000134   0.157113     1.0
2   0.000149   0.157178     1.0
3   0.000149   0.157243     1.0
4   0.000153   0.157308     1.0

In [99]:
pr_df.iloc[::50]

threshold  precision    recall
0      0.000057   0.157049  1.000000
50     0.000339   0.160354  1.000000
100    0.000459   0.163801  1.000000
150    0.000530   0.167399  1.000000
200    0.000594   0.171159  1.000000
250    0.000651   0.175092  1.000000
300    0.000713   0.179210  1.000000
350    0.000790   0.183526  1.000000
400    0.000856   0.188055  1.000000
450    0.000950   0.192814  1.000000
500    0.001040   0.197819  1.000000
550    0.001129   0.203092  1.000000
600    0.001234   0.208653  1.000000
650    0.001324   0.214527  1.000000
700    0.001480   0.220742  1.000000
750    0.001625   0.227327  1.000000
800    0.001858   0.234317  1.000000
850    0.002174   0.241751  1.000000
900    0.002847   0.249672  1.000000
950    0.005759   0.258130  1.000000
1000   0.058756   0.265778  0.994751
1050   0.070276   0.273256  0.986877
1100   0.081711   0.280543  0.976378
1150   0.091040   0.289969  0.971129
1200   0.104750   0.300979  0.968504
1250   0.115045   0.312925  0.965879
1300   0.125289   0.324156  0.958005
1350   0.134498   0.335502  0.947507
1400   0.146713   0.349903  0.942257
1450   0.157304   0.365779  0.937008
1500   0.170562   0.381210  0.926509
1550   0.183765   0.401826  0.923885
1600   0.196867   0.424939  0.921260
1650   0.214777   0.448454  0.913386
1700   0.237099   0.473829  0.902887
1750   0.256540   0.504438  0.895013
1800   0.280128   0.541534  0.889764
1850   0.308533   0.585069  0.884514
1900   0.338889   0.636882  0.879265
1950   0.387902   0.697479  0.871391
2000   0.459048   0.776995  0.868766
2050   0.925410   0.840426  0.829396
2100   0.951638   0.840491  0.719160
2150   0.958787   0.836957  0.606299
2200   0.966030   0.845133  0.501312
2250   0.970652   0.863636  0.398950
2300   0.976245   0.857143  0.283465
2350   0.981633   0.855263  0.170604
2400   0.986865   0.769231  0.052493

In [100]:
best_threshold = 0.17

y_pred_custom = (y_prob >= best_threshold).astype(int)

print("Accuracy:", accuracy_score(y_test, y_pred_custom))
print("Recall:", recall_score(y_test, y_pred_custom))
confusion_matrix(y_test, y_pred_custom)

Accuracy: 0.7522671063478977
Recall: 0.926509186351706


array([[1472,  573],
       [  28,  353]])

Reajuste de modelo

In [121]:
t = 0.40
y_pred_40 = (y_proba >= t).astype(int)

cm_40 = confusion_matrix(y_test, y_pred_40)
prec_40 = precision_score(y_test, y_pred_40)
rec_40 = recall_score(y_test, y_pred_40)

print("Threshold:", t)
print("Precision:", round(prec_40, 4))
print("Recall:", round(rec_40, 4))
print("Matriz de confusión:\n", cm_40)

print("Clientes en riesgo:", cm_40[0,1] + cm_40[1,1])
print("Porcentaje base:", (cm_40[0,1] + cm_40[1,1]) / len(y_test))

Threshold: 0.4
Precision: 0.7043
Recall: 0.8688
Matriz de confusión:
 [[1906  139]
 [  50  331]]
Clientes en riesgo: 470
Porcentaje base: 0.19373454245671887


Métricas

Recall = 331 / (331 + 50) = 0.87

Clientes en riesgo = 139 + 331 = 470

% base ≈ 19%

Se detecta el 87% del churn
Con solo 19% de la base impactada

Buen balance costo–beneficio.

#Serialización

In [123]:
import joblib

joblib.dump(model, "modelo_regresion_logistica.pkl")

['modelo_regresion_logistica.pkl']

Carga del modelo

In [124]:
modelo_cargado = joblib.load("modelo_regresion_logistica.pkl")

probas_nuevas = modelo_cargado.predict_proba(X_test)[:, 1]
pred_nuevas = (probas_nuevas >= 0.40).astype(int)

confusion_matrix(y_test, pred_nuevas)

array([[1906,  139],
       [  50,  331]])